CRAYSAT-1530: Document `sat bootprep` improvements (#29)
Cray-HPE/docs-sat: docs/sat_bootprep.md

These sections may be specified in any order, and any of the sections may be
omitted if desired.

### Providing a schema version

The `sat bootprep` input file is validated against a versioned schema
definition. The input file should specify the version of the schema with which
it is compatible under a `schema_version` key. For example:

```yaml
---
schema_version: 1.0.2
```

The current `sat bootprep` input file schema version can be viewed with the
following command:

```screen
ncn-m001# sat bootprep view-schema | grep '^version:'
version: '1.0.2'
```

The `sat bootprep run` command validates the schema version specified
in the input file. The command also makes sure that the schema version
of the input file is compatible with the schema version understood by the
current version of `sat bootprep`. For more information on schema version
validation, see the `schema_version` property description in the bootprep
input file schema. For more information on viewing the bootprep input file
schema in either raw form or user-friendly HTML form, see [Viewing the exact
schema specification](#viewing-the-exact-schema-specification) or

> [Generating user-friendly documentation](#generating-user-friendly-documentation).

Author: This one renders okay. As discussed in scrum, I think it's okay to have the hard line breaks (newline characters) in the link text, but it is not okay to introduce extra whitespace (space, tabs, newlines) in between the closing square bracket `]` and the opening parenthesis `(`. It's also not okay to have whitespace in the link destination.

Here's a reference to the GitHub-flavored markdown spec: https://github.github.com/gfm/#links

The default `sat bootprep` input files provided by the `hpc-csm-software-recipe`
release distribution already contain the correct schema version.

### Defining CFS configurations

The CFS configurations are defined under a `configurations` key. Under this

> key, the user can list one or more configurations to create. For each
> configuration, a name should be given in addition to the list of layers which

Reviewer: ```suggestion
key, you can list one or more configurations to create. For each
configuration, give a name in addition to the list of layers that
```

contain a `name` key and a `base` key.

The `name` key defines the name of the resulting IMS image. The `base` key
defines the base image to be configured or the base recipe to be built and
optionally configured. One of the following keys must be present under the
`base` key:

- Use an `ims` key to specify an existing image or recipe in IMS.
- Use a `product` key to specify an image or recipe provided by a
  particular version of a product. Note that this is only possible if the
  product provides a single image or recipe.
- Use an `image_ref` key to specify another image from the input file
  using its `ref_name`.

Images may also contain the following keys:

- Use a `configuration` key to specify a CFS configuration with which to
customize the built image. If a configuration is specified, then configuration
groups must also be specified using the `configuration_group_names` key.
- Use a `ref_name` key to specify a unique name that can refer to this image
within the input file in other images or in session templates. The `ref_name`
key allows references to images from the input file that have dynamically
generated names as described in [Dynamic variable substitutions]

> (#dynamic-variable-substitutions).

Author: Good catch, this one was broken.

> (#dynamic-variable-substitutions).

Author: I'm also noticing now that I should fix up some of the spacing on these multi-line list items. I think our style is to have the start of the text line up on each subsequent line of a list item, so some of these lines should be indented by 2 spaces. I'll make that change as well.

and `bos_parameters` keys. The `name` key defines the name of the resulting BOS
session template.  The `image` key defines the image to use in the BOS session
template. One of the following keys must be present under the `image` key:

- Use an `ims` key to specify an existing image or recipe in IMS.
- Use an `image_ref` key to specify another image from the input file
  using its `ref_name`.

The `configuration` key defines the CFS configuration specified
in the BOS session template.

The `bos_parameters` key defines parameters that are passed through directly to
the BOS session template. The `bos_parameters` key should contain a `boot_sets`
key, and each boot set in the session template should be specified under
`boot_sets`. Each boot set can contain the following keys, all of
which are optional:

- `kernel_parameters`: the parameters passed to the kernel on the command line
- `network`: the network over which the nodes will boot
- `node_list`: nodes to add to the boot set
- `node_roles_groups`: HSM roles to add to the boot set
- `node_groups`: HSM groups to add to the boot set
- `rootfs_provider`: the root file system provider

> - `rootfs_provider_passthrough`: parameters to add to the `rootfs=` kernel

Author: Huh, I wonder what happened here. Might have been something to do with applying suggestions. Good catch, I'll fix this up.

- Use a `kernel_parameters` key to specify the parameters passed to the kernel on the command line.
- Use a `network` key to specify the network over which the nodes boot.
- Use a `node_list` key to specify the nodes to add to the boot set.
- Use a `node_roles_groups` key to specify the HSM roles to add to the boot set.
- Use a `node_groups` key to specify the HSM groups to add to the boot set.
- Use a `rootfs_provider` key to specify the root file system provider.
- Use a `rootfs_provider_passthrough` key to specify the parameters to add to the `rootfs=`
kernel parameter.

As mentioned above, the parameters under `bos_parameters` are passed through
directly to BOS. For more information on the properties of a BOS boot set,
refer to the section **BOS Session Templates** in the [*Cray

> System Management Documentation*](https://cray-hpe.github.io/docs-csm/).

Author: This one's okay as well since the newline occurs in the link title.

expression, it must be quoted to pass YAML syntax checking.

Jinja2 expressions can also use filters and Python's built-in string methods to
manipulate the variable values. For example, suppose only the major and minor
components of a COS version are to be used in the branch name for the COS
layer of the CFS configuration. You can use the `split` string method to
achieve this as follows:

```yaml
---
configurations:
- name: compute-{{recipe.version}}
  layers:
  - name: cos-compute-integration-{{cos.version}}
    playbook: cos-compute.yaml
    product:
      name: cos
      version: "{{cos.version}}"
      branch: integration-{{cos.version.split('.')[0]}}-{{cos.version.split('.')[1]}}
```

### Dynamic variable substitutions

In addition to the product version variables provided by the HPC CSM Software
Recipe as described in [HPC CSM Software Recipe variable
substitutions](#hpc-csm-software-recipe-variable-substitutions), additional
variables are available. These variables are dynamic in that their values are
determined at run-time based on the context in which they appear. The additional
variables are as follows:

> 

Author: Seems like another suggestion application error. Good catch.

Additional variables are available besides the product version variables
provided by the HPC CSM Software Recipe. (See [HPC CSM Software Recipe variable
substitutions](#hpc-csm-software-recipe-variable-substitutions) for more

> information.) These additional variables are dynamic because their values are

Author: This one's okay as-is also

      name: cpe
      version: "{{cpe.version}}"
      branch: integration-{{cpe.version}}

images:
- name: "{{base.name}}"
  ref_name: base_cos_image
  base:
    product:
      name: cos
      type: recipe
      version: "{{cos.version}}"

- name: compute-{{base.name}}
  ref_name: compute_image
  base:
    image_ref: base_cos_image
  configuration: compute-{{recipe.version}}
  configuration_group_names:
  - Compute

session_templates:
- name: compute-{{recipe.version}}
  image:
    image_ref: compute_image
  configuration: compute-{{recipe.version}}
  bos_parameters:
    boot_sets:
      compute:
        kernel_parameters: ip=dhcp quiet spire_join_token=${SPIRE_JOIN_TOKEN}
        node_roles_groups:
        - Compute
        rootfs_provider_passthrough: "dvs:api-gw-service-nmn.local:300:hsn0,nmn0:0"
```

### Accessing default bootprep input files

Default bootprep input files are delivered by the HPC CSM Software Recipe
product. You can access these files by cloning the `hpc-csm-software-recipe`
repository.

To do this, follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
variables](#viewing-hpc-csm-software-recipe-variables). Then, access the files in the

> `bootprep` directory of that repository:

Author: This one's okay as-is also.
